plugins/experimental/wasm/ats_wasm.cc

```diff
@@ -0,0 +1,149 @@
+/*
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+#include "ats_wasm.h"
+
+#include <string>
+
+namespace proxy_wasm
+{
+// extended constructors to initialize mutex
+Wasm::Wasm(const std::shared_ptr<WasmHandleBase> &base_wasm_handle, const WasmVmFactory &factory)
+  : WasmBase(base_wasm_handle, factory)
+{
+  mutex_ = TSMutexCreate();
+}
+
+Wasm::Wasm(std::unique_ptr<WasmVm> wasm_vm, std::string_view vm_id, std::string_view vm_configuration, std::string_view vm_key,
+           std::unordered_map<std::string, std::string> envs, AllowedCapabilitiesMap allowed_capabilities)
+  : WasmBase(std::move(wasm_vm), vm_id, vm_configuration, vm_key, envs, allowed_capabilities)
+{
+  mutex_ = TSMutexCreate();
+}
+
+// function to retrieve mutex
+TSMutex
+Wasm::mutex() const
+{
+  return mutex_;
+}
+
+// functions to create contexts
+ContextBase *
+Wasm::createVmContext()
+{
+  return new Context(this);
+}
+
+ContextBase *
+Wasm::createRootContext(const std::shared_ptr<PluginBase> &plugin)
+{
+  return new Context(this, plugin);
+}
+
+ContextBase *
+Wasm::createContext(const std::shared_ptr<PluginBase> &plugin)
+{
+  return new Context(this, plugin);
+}
+
+// Function to start a new root context
+Context *
+Wasm::start(std::shared_ptr<PluginBase> plugin, TSCont contp)
+{
+  auto it = root_contexts_.find(plugin->key());
+  if (it != root_contexts_.end()) {
+    Context *c = (Context *)it->second.get();
+    if (c->scheduler_cont() == nullptr) {
+      c->initialize(contp);
+    } else {
+      // keep the old continuation
+      TSContDestroy(contp);
+    }
+    c->onStart(plugin);
+    return c;
+  }
+  auto context     = std::unique_ptr<Context>((Context *)createRootContext(plugin));
```

Comment:
This is a strange use of unique_ptr.  context is moved to root_contexts_ and also returned as the pointer.  Perhaps this should be shared (or weak_shared)?  Since `createRootContext` just calls new anyway, this code could be cleaner.

Reply:
Also preferable to avoid C-style casting.

Reply:
root_contexts_ from the base class needs a unique_ptr as the value